[BS2-21] BE: Withdraw balance to bank

tapabratadey opened the pull request:
1. Withdraw button in user Profile page w/ a balance

2. Balance shows on the user profile, when they click withdraw (if balance > 0) then we’re imitating a transfer from JumpStarter to their bank account, so basically setting the balance to 0.

bennettl commented on src/backend/controller/user/user.ts:
```diff
@@ -3,28 +3,120 @@ import {
+    // break down request <===========
+    const { email } = req.user;
+    // call service <===============
+    const userData = await getUser(email);
```
the service method should throw an exception if user is not found

tapabratadey replied:
Thanks, fixed it!

bennettl commented on src/backend/controller/user/user.ts:
```diff
@@ -3,28 +3,120 @@ import {
+        userData,
+      });
+    } else {
+      res.status(404).json({
```
No need for else statement if exception is throw and caught in `catch`

tapabratadey replied:
Thanks, fixed it, sending 500 from the handler instead.

tapabratadey commented on src/backend/controller/user/user.ts:
```diff
@@ -3,28 +3,120 @@ import {
+    }
+  } catch (error) {
+    console.log('ERROR @getUserController in controller/user/user.ts', error);
+    res.status(500).json(error.message);
```
Thanks, using `http status codes` package now

bennettl commented on src/backend/controller/user/user.ts:
```diff
@@ -1,30 +1,47 @@
-  insertUser(req, res);
-};
+  post: async (req: Request) => {
+    const firstName = req.body.firstName;
```
Use destructing const { body: {firstName, lastName, email}} = req;

tapabratadey replied:
Resolved

bennettl commented on src/backend/controller/user/user.ts:
```diff
@@ -1,30 +1,47 @@
+    const userData = await UserDbService.update(
+      firstName,
+      lastName,
+      bio,
```
Let's create only one parameter, it doesn't scale well for large objects to have X number of params per attribute

tapabratadey replied:
Resolved

bennettl commented on src/backend/controller/user/user.ts:
```diff
@@ -1,30 +1,47 @@
+
+  delete: async (req: Request) => {
+    const email = req.user.email;
+    const userData = await UserDbService.delete(email);
```
typically we delete by user id

tapabratadey replied:
Resolved. Using firebase `uid` now, created a column in User table which stores user `uid` during sign up and is referred during each api call.

bennettl commented on src/backend/middleware/verify_request.ts:
```diff
@@ -12,7 +19,9 @@ export const verifyRequest = (handler) => {
-      return res.status(401).send('Access token is required');
+      return res
+        .status(StatusCodes.UNAUTHORIZED)
+        .send('Access token is required');
```
We should have a generic error JSON object instead of a string

bennettl commented on src/backend/services/db/user/user_db.ts:
```diff
@@ -1,127 +1,114 @@
+        .into(User)
+        .values([
+          {
+            firstName: firstName,
```
` firstName: firstName,` is not necessary in javascript, you can just use ` firstName`

tapabratadey replied:
Resolved

bennettl commented on src/pages/api/users/get.ts:
```diff
@@ -9,16 +16,26 @@ const handler = async (req: Request, res: NextApiResponse) => {
+            userData,
+          });
+        } else {
+          res.status(StatusCodes.NOT_FOUND).json({
```
Error cases should not be handled with if / else, it should be handled with errors and try / catch

tapabratadey replied:
Resolved. Service returns data or throws an Error which gets sent down to the handler as a promise. try / catch in handler makes the decision on how to respond to the client.

bennettl commented on src/pages/api/users/payout.ts:
```diff
@@ -0,0 +1,42 @@
+        }
+        break;
+      default:
+        res.status(StatusCodes.METHOD_NOT_ALLOWED).json({
```
You are mixing string literals in order parts of the code and JSON objects for error responses, we need to be consistent with structure of response type or else frontend will have a hard time processing it

bennettl commented on src/pages/app/profile/index.tsx:
```diff
@@ -40,7 +41,32 @@ const userData2 = [
+  const [isMoneyTransferred, setMoneyTransferred] = useState(false);
+
+  const handlePayOut = async () => {
+    const headers = {
```
Let's create a generic axios instance with interceptor for these generic http headers

tapabratadey replied:
Awesome advice! Didn't know axios is so powerful :p Setup axios instance and response interceptor

tapabratadey commented on src/backend/controller/user/user.ts:
```diff
@@ -1,30 +1,47 @@
+    const lastName = req.body.lastName;
+    const bio = req.body.bio;
+    const avatar = req.body.avatar;
+    const userData = await UserDbService.update(
```
Resolved

tapabratadey commented:
Learned a lot about best practices and error handling, thanks for such a detailed review!

Updates:
1. Refactored code to adapt to architecture pattern.

- Service layer throws an error if there's an error or else throws a promise
- The controller forwards the promise to the handler
- Handler catches an error if there's any, logs and responds to the client with errors

3. Adding Error classes in `src/helpers/ErrorHandling/error.ts`
4. Implemented Axios instance/response intercepts in `src/axios/instance.ts`

- Cleaned up exposed API URL, headers, and responses.

Thanks again!
